src/services/paras/ParasService.ts

```diff
@@ -0,0 +1,441 @@
+	 *
+	 * @param hash `BlockHash` to make call at
+	 * @returns all the current registered paraIds and their lifecycle status
+	 */
+	async paras(hash: BlockHash): Promise<IParas> {
+		const [{ number }, paraLifecycles] = await Promise.all([
+			this.api.rpc.chain.getHeader(hash),
+			this.api.query.paras.paraLifecycles.entriesAt<ParaLifecycle, [ParaId]>(
+				hash
+			),
+		]);
+
+		const parasPromises = paraLifecycles.map(async ([k, paraLifecycle]) => {
+			const paraId = k.args[0];
+			let onboardingAs: ParaType | undefined;
+			if (paraLifecycle.isOnboarding) {
+				const paraGenesisArgs = await this.api.query.paras.paraGenesisArgs.at<ParaGenesisArgs>(
+					hash,
+					paraId
+				);
+				onboardingAs = paraGenesisArgs.parachain.isTrue
+					? 'parachain'
+					: 'parathread';
+			}
+
+			return {
+				paraId,
+				paraLifecycle,
+				onboardingAs,
+			};
+		});
+
+		return {
+			at: {
+				hash,
+				height: number.unwrap().toString(10),
+			},
+			paras: await Promise.all(parasPromises),
+		};
+	}
+
+	/**
+	 * Calculate the current lease period index.
+	 *
+	 * @param blockHeight current blockheight
+	 * @param leasePeriod duration of lease period
+	 */
+	private currentLeasePeriodIndex(now: BN): BN {
+		const leasePeriod = this.api.consts.slots.leasePeriod as BlockNumber;
+		return now.div(leasePeriod);
+	}
+
+	/**
+	 * The offset into the ending samples of the auction. When we are not in the
+	 * ending phase of the auction we can use 0 as the offset, but we do not return
+	 * that here in order to closely mimic `Auctioneer::is_ending` impl in
+	 * polkadot's `runtime::common::auctions`.
+	 *
+	 * @param now current block number
+	 * @param beginEnd block number of the start of the auction's ending period
+	 */
+	private endingOffset(now: BN, beginEnd: IOption<BN>): IOption<BN> {
+		if (!isSome(beginEnd)) {
+			return null;
+		}
+
+		const afterEarlyEnd = now.sub(beginEnd);
+		if (afterEarlyEnd.lten(0)) {
+			return null;
+		}
+
+		// Once https://github.com/paritytech/polkadot/pull/2848 is merged no longer
+		// need a fallback of 1
+		const sampleLength =
+			(this.api.consts.auctions.sampleLength as BlockNumber) || new BN(1);
+		return afterEarlyEnd.div(sampleLength);
+	}
+
+	/**
+	 * Enumerate in order all the lease sets (SlotRange expressed as a set of
+	 * lease periods) that an `auctions::winning` array covers.
+	 *
+	 * @param leasePeriodIndex
+	 */
+	private enumerateLeaseSets(leasePeriodIndex: BN): number[][] {
+		const leasePeriodIndexNumber = leasePeriodIndex.toNumber();
+		const lPPS =
+			(this.api.consts.auctions.leasePeriodsPerSlot as u32)?.toNumber() ||
+			LEASE_PERIODS_PER_SLOT_FALLBACK;
+
+		const ranges: number[][] = [];
+		for (let start = 0; start < lPPS; start += 1) {
+			for (let end = start; end < lPPS; end += 1) {
+				const slotRange = [];
+				for (let i = start; i <= end; i += 1) {
+					slotRange.push(i + leasePeriodIndexNumber);
+				}
+				ranges.push(slotRange);
+			}
+		}
```

Comment:
The third inner loop builds the sub arrays that represent the `SlotRange` variant. You can think of the outer two loops as creating the start and the end of the range, then the inner most loop iterates through that start and end to build it for us.
 If we have 3 lease periods per slot (`lPPS`), the outer loop start at 0 and the 2nd inner loop builds increasing ranges starting from 0:
`[0], [0, 1], [0, 1, 2]`. 
Once we have built our 0 starting ranges we then increment the outermost loop and start building our 1 starting ranges:
`[1], [1, 2]`
And finally we increment the outer most loop to 2, building our two starting ranges
`[2]`

Put all those ranges together in the order they where produced and we get:
`[[0], [0, 1], [0, 1, 2], [1], [1, 2], [2]]`
So now we have an array, where each index corresponds to the same `SlotRange` that would be at that index in the `winning` array.

Reply:
You have other ideas for generating the same sets in this order I am open to it, this is just the first thing I could think of and was not recursive.

Reply:
I see, ty for laying it out. It's fine as-is; maybe it could be a `fold` to reduce the cyclometric complexity a bit, but it'd still require the reader to think a bit so not much to gain there. Perhaps toss in a comment with an example of expected output?

Reply:
TIL: cyclometric complexity

Reply:
I added an explanation comment with expected output for an input.

How would you use a `fold` (in js I believe its `reduce`) to reduce complexity? I can see how to replace any of the for loops with a `fold`, but can't think of a way to do it that reduces any sort of complexity. (I probably don't understand cyclometric (cyclomatic?) complexity)

Reply:
I'd have to tinker with it, I don't know that it can be done (or that it'd be better, really), just a hunch that there are combinators in e.g. lodash or itertools that my gut tells me should be able to do this. But it's totally not important. (And yes, cyclomatic complexity, my bad!)